Assets/MRTK/Providers/WindowsMixedReality/XRSDK/WindowsMixedRealityDeviceManager.cs

```diff
@@ -126,9 +140,96 @@ public override bool CheckCapability(MixedRealityCapability capability)
+        /// </summary>
+        private readonly Dictionary<uint, MotionControllerState> trackedMotionControllerStates = new Dictionary<uint, MotionControllerState>();
+
+        private readonly Dictionary<uint, GenericXRSDKController> activeMotionControllers = new Dictionary<uint, GenericXRSDKController>();
+#endif
+
         #endregion IMixedRealityCapabilityCheck Implementation
 
+
         #region Controller Utilities
+#if HP_CONTROLLER_ENABLED
+        private static readonly ProfilerMarker GetOrAddControllerPerfMarker = new ProfilerMarker("[MRTK] WindwosMixedRealityXRSDKDeviceManager.GetOrAddController");
+
+        protected override GenericXRSDKController GetOrAddController(InputDevice inputDevice)
+        {
+            using (GetOrAddControllerPerfMarker.Auto())
+            {
+                GenericXRSDKController detectedController = base.GetOrAddController(inputDevice);
+                SupportedControllerType currentControllerType = GetCurrentControllerType(inputDevice);
+
+                // Add the Motion Controller state if it's an HPMotionController
+                if (currentControllerType == SupportedControllerType.HPMotionController)
+                {
+                    lock (trackedMotionControllerStates)
+                    {
+                        uint controllerId = GetControllerId(inputDevice);
+                        if (trackedMotionControllerStates.ContainsKey(controllerId))
+                        {
+                            ((HPMotionController)detectedController).MotionControllerState = trackedMotionControllerStates[controllerId];
+                        }
+                    }
+                }
+
+                return detectedController;
+            }
+        }
+
+        private void AddTrackedMotionController(object sender, MotionController motionController)
+        {
+            lock (trackedMotionControllerStates)
+            {
+                uint controllerId = GetControllerId(motionController);
+                trackedMotionControllerStates[controllerId] = new MotionControllerState(motionController);
+
+                if (activeMotionControllers.ContainsKey(controllerId) && activeMotionControllers[controllerId] is HPMotionController hpController)
+                {
+                    hpController.MotionControllerState = trackedMotionControllerStates[controllerId];
+                }
+            }
+        }
+
+        private void RemoveTrackedMotionController(object sender, MotionController motionController)
+        {
+            lock (trackedMotionControllerStates)
+            {
+                uint controllerId = GetControllerId(motionController);
+                trackedMotionControllerStates.Remove(controllerId);
+
+                if (activeMotionControllers.ContainsKey(controllerId))
+                {
+                    ((HPMotionController)activeMotionControllers[controllerId]).MotionControllerState = null;
```

Comment:
Looks like this is using the old pattern instead of the one you just updated on line 192

```c#
 && activeMotionControllers[controllerId] is HPMotionController hpController
```